File: 042. 接雨水.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        result = 0
        left = 0
        right = len(height)-1
        max_left = height[left]
        max_right = height[right]
        while left<right:
            max_left = max(max_left,height[left])
            max_right = max(max_right,height[right])
            if max_left<max_right:
                result += max_left - height[left]
                left+=1
            else:
                result += max_right - height[right]
                right-=1
        return result
s = Solution()
print(s.trap([0,1,0,2,1,0,1,3,2,1,2,1]))

File: 042. 接雨水.go
package main

func trap(height []int) int {
	min := func(x, y int) int {
		if x < y {
			return x
		}
		return y
	}
	max := func(x, y int) int {
		if x > y {
			return x
		}
		return y
	}

	n := len(height)
	tallestLeft, tallestRight := make([]int, n), make([]int, n)
	res := 0

	for i := 1; i < n; i++ {
		tallestLeft[i] = max(tallestLeft[i-1], height[i-1]) //不是 height[i]
	}
	for i := n - 2; i >= 0; i-- {
		tallestRight[i] = max(tallestRight[i+1], height[i+1])
	}

	for i := 0; i < n; i++ {
		res += max(0, min(tallestLeft[i], tallestRight[i])-height[i])
	}

	return res
}
